R/assertions.R

```diff
@@ -27,16 +39,15 @@ assert_if <- function(data, expr, description = NA, obligatory = FALSE, skip_cha
 }
 
 #' @export
-assert_cols <- function(data, ..., predicate = NULL, predicate_generator = NULL, obligatory = FALSE, description = NA,
+assert_cols <- function(data, predicate, ..., obligatory = FALSE, description = NA,
                         skip_chain_opts = FALSE, success_fun = assertr::success_append, error_fun = assertr::error_append,
                         defect_fun = assertr::defect_append) {
 
-  assertr_function <- if (!is.null(predicate_generator)) assertr::insist else assertr::assert
-  predicate_arg <- if (!is.null(predicate_generator)) predicate_generator else predicate
+  assertr_function <- get_assert_method(predicate, list(direct = assertr::assert, generator = assertr::insist))
```

Comment:
Do we really need to provide `methods` argument? Are there any other predicate types possible? Maybe this is a premature optimization that causes confusion. I think we just need a function like `get_predicate_type`, for example:

```
get_assert_method <- function(predicate) {
  predicate_output <- predicate(0:1)
  if (is.logical(predicate_output)) {
    return(assertr::assert)
  } else if (is.function(predicate_output) && is.logical(predicate_output(0:1))) {
    return(assertr::insist)
  }  else {
    stop("Predicate should be a function returning logical vector or function")
  }
}
```

WDYT?

Reply:
Your example covers only 50% of cases. I've added this because we also have assert_rows and insist_rows case.

Reply:
I see! I missed that this function is also used in `assert_rows` with different methods. So, if we assume anyway that there is "direct" and "generator" method only, can we have a single function like this:

```
get_assert_method <- function(predicate, method = list(direct = assertr::assert, generator = assertr::insist)) {
  predicate_output <- predicate(0:1)
  if (is.logical(predicate_output)) {
    return(method$direct)
  } else if (is.function(predicate_output) && is.logical(predicate_output(0:1))) {
    return(method$generator)
  }  else {
    stop("Predicate should be a function returning logical vector or function")
  }
}
```

Having such default argument for `methods` in function interface clarifies me a lot. WDYT now?

Reply:
fixed